data_structures/arrays/first_missing_positive.py

```diff
@@ -0,0 +1,60 @@
+def first_missing_positive(nums: list[int]) -> int:
+    """
+    Given an unsorted integer array nums, returns the smallest missing positive integer.
+
+    The algorithm must run in O(n) time and use O(1) auxiliary space.
+
+    Hints:
+    - Try to place each number in its correct position, so that nums[i] equals i + 1.
+    - Iterate through the array to find the first position where nums[i] != i + 1,
+      indicating the missing positive integer.
+
+    :param nums: The input unsorted integer array.
+    :return: The smallest missing positive integer.
+
+    Examples:
+    >>> first_missing_positive([1, 2, 0])
```

Comment:
Please add doctests for
```
>>> first_missing_positive([])
>>> first_missing_positive([0])
>>> first_missing_positive([1])
>>> first_missing_positive([1.1, 2.2, 3.3])
>>> first_missing_positive([-1, -2, -3])
```

Reply:
thanks for the feedback .. i have done it